File: 0695-max-area-of-island/0695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:

        rows: int = len(grid)
        cols: int = len(grid[0])

        directions: List[Tuple[int, int]] = [(0, 1), (1, 0), (-1, 0), (0, -1)]

        def isValid(row: int, col: int) -> None:
            return 0 <= row < rows and 0 <= col < cols

        def dfs(row: int, col: int) -> bool:
            if not isValid(row, col) or grid[row][col] != 1:
                return 0

            grid[row][col] = -1
            
            area = 1

            for dr, dc in directions:
                area += dfs(row + dr, col + dc)
            
            return area
        
        max_area: int = 0
        for row in range(rows):
            for col in range(cols):
                if grid[row][col] == 1:
                    area = dfs(row, col)
                    max_area = max(max_area, area)
                
        return max_area





        

File: 0695-max-area-of-island/0695-max-area-of-island.java
class Solution {

    private static final int[][] directions = new int[][] {
        {0, 1},
        {1, 0},
        {0, -1},
        {-1, 0}
    };
    
    public int maxAreaOfIsland(int[][] grid) {

        int rows = grid.length;
        int cols = grid[0].length;

        int max_area = 0;
        for(int row = 0; row < rows; row++){
            for(int col = 0; col < cols; col++){
                if (grid[row][col] == 1){
                    int area = dfs(grid, row, col, rows, cols);
                    max_area = Math.max(area, max_area);
                }
            }
        }

        return max_area;
    }

    private boolean isValid(int[][] grid, int row, int col, int rows, int cols) {
        return (row >= 0 && row < rows) && (col >= 0 && col < cols);
    };

    int dfs(int[][] grid, int row, int col, int rows, int cols) {

        if (!isValid(grid, row, col, rows, cols) || grid[row][col] != 1){
            return 0;
        }

        int area = 1;
        grid[row][col] = -1;

        for (int[] direction: directions) {
            int r = direction[0] + row;
            int c = direction[1] + col;

            area += dfs(grid, r, c, rows, cols);
        }

        return area;
    };
}